### PR TITLE
[Donut] Legends multi selection for Donut Charts

### PR DESCRIPTION
**Pull request by srmukher**
Legends multi selection for Donut Charts
image

### CODE REVIEW

File: packages/charts/react-charting/src/components/DonutChart/Arc/Arc.tsx [hunk 1]
```diff
@@ -31,12 +31,11 @@ export class Arc extends React.Component<IArcProps, IArcState> {
   }
```
Comment (fabricteam) [inline packages/charts/react-charting/src/components/DonutChart/Arc/Arc.tsx:31]: ## 🕵🏾‍♀️ [<b data-uid='vrt-gist'></b> visual regressions to review in the fluentuiv8 Visual Regression Report](https://aka.ms/visual-changes-review?organization=uifabric&project=fabricpublic&prId=33447&commitId=bc2f7abb351b1cf667150b475fe17194caadb014&env=prod&clientType=FLUENTUI)


  <details><summary><b>Callout</b> 5 screenshots</summary>

|Image Name|Diff(in Pixels)| Image Type
|-|-|-|
|Callout.Rendering callout attached to a rectangle.chromium.png|1493| Changed|
|Callout.Top right edge.chromium.png|1594| Changed|
|Callout.Left center.chromium.png|2616| Changed|
|Callout.Root.chromium.png|2309| Changed|
|Callout.Right bottom edge.chromium.png|3005| Changed|

</details>


  <details><summary><b>react-charting-HeatMapChart</b> 1 screenshots</summary>

|Image Name|Diff(in Pixels)| Image Type
|-|-|-|
|react-charting-HeatMapChart.Basic RTL.chromium.png|372| Changed|

</details>


  <details><summary><b>react-charting-LineChart</b> 1 screenshots</summary>

|Image Name|Diff(in Pixels)| Image Type
|-|-|-|
|react-charting-LineChart.Gaps.chromium.png|1| Changed|

</details>




<div id="vrtCommentfluent-ui_VRT_Pipeline_v8"/>

File: packages/react-examples/src/react-charting/DonutChart/DonutChart.Multiple.Example.tsx [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+  DataVizGradientPalette,
+} from '@fluentui/react-charting';
+import { Toggle } from '@fluentui/react/lib/Toggle';
+
+interface IDonutChartState {
+  enableGradient: boolean;
+  roundCorners: boolean;
+}
+
+export class DonutChartMultipleExample extends React.Component<IDonutChartProps, IDonutChartState> {
```
Comment (AtishayMsft) [inline packages/react-examples/src/react-charting/DonutChart/DonutChart.Multiple.Example.tsx:19]: >DonutChartMultipleExample [](http://example.com/codeflow?start=13&length=25)

don't create a new example. Add a toggle in the existing example.

File: packages/charts/react-charting/src/components/DonutChart/Arc/Arc.tsx [hunk 3]
```diff
@@ -31,12 +31,11 @@ export class Arc extends React.Component<IArcProps, IArcState> {
...
 
   public render(): JSX.Element {
-    const { arc, href, focusedArcId } = this.props;
+    const { arc, href, focusedArcId, activeArc } = this.props;
     const getClassNames = classNamesFunction<IArcStyleProps, IArcStyles>();
     const id = this.props.uniqText! + this.props.data!.data.legend!.replace(/\s+/, '') + this.props.data!.data.data;
-    const opacity: number =
-      this.props.activeArc === this.props.data!.data.legend || this.props.activeArc === '' ? 1 : 0.1;
-
+    const isActive = activeArc?.includes(this.props.data?.data.legend!);
```
Comment (AtishayMsft) [inline packages/charts/react-charting/src/components/DonutChart/Arc/Arc.tsx:37]: >const isActive  [](http://example.com/codeflow?start=4&length=15)

why repeating this logic.

File: packages/charts/react-charting/src/components/DonutChart/DonutChart.base.tsx [hunk 4]
```diff
@@ -354,7 +374,7 @@ export class DonutChartBase extends React.Component<IDonutChartProps, IDonutChar
    * Note: This won't work in case of multiple legends selection.
    */
   private _getHighlightedLegend() {
-    return this.state.selectedLegend || this.state.activeLegend;
```
Comment (AtishayMsft) [inline packages/charts/react-charting/src/components/DonutChart/DonutChart.base.tsx:357]: > return this.state.selectedLegend || this.state.activeLegend; [](http://example.com/codeflow?start=3&length=61)

This is an or condition between selected and active.

File: packages/charts/react-charting/src/components/DonutChart/DonutChart.base.tsx [hunk 5]
```diff
@@ -324,15 +338,21 @@ export class DonutChartBase extends React.Component<IDonutChartProps, IDonutChar
...
-    if (valueInsideDonut !== undefined && (highlightedLegend !== '' || this.state.showHover)) {
+    const highlightedLegends = this._getHighlightedLegend();
+    if (valueInsideDonut !== undefined && (highlightedLegends.length > 0 || this.state.showHover)) {
       let legendValue = valueInsideDonut;
+      let totalValue = 0;
       data!.map((point: IChartDataPoint, index: number) => {
-        if (point.legend === highlightedLegend || (this.state.showHover && point.legend === this.state.legend)) {
+        if (
+          highlightedLegends.includes(point.legend!) ||
+          (this.state.showHover && point.legend === this.state.legend)
```
Comment (AtishayMsft) [inline packages/charts/react-charting/src/components/DonutChart/DonutChart.base.tsx:348]: > point.legend === this.state.legend [](http://example.com/codeflow?start=34&length=35)

you are adding selected legend and hover legend values together?

File: packages/charts/react-charting/src/components/DonutChart/Arc/Arc.tsx (hunk 3 above)
Comment (srmukher) [reply packages/charts/react-charting/src/components/DonutChart/Arc/Arc.tsx:37]: so that it can be used to calculate both opacity and data-is-focusable without the need to add same checks again

File: packages/charts/react-charting/src/components/DonutChart/DonutChart.base.tsx (hunk 4 above)
Comment (srmukher) [reply packages/charts/react-charting/src/components/DonutChart/DonutChart.base.tsx:357]: For multiple select also checking if this.state.selectedLegends is populated other return the this.state.activeLegend
    return this.state.selectedLegends.length > 0 ? this.state.selectedLegends : [this.state.activeLegend];

File: packages/react-examples/src/react-charting/DonutChart/DonutChart.Multiple.Example.tsx (hunk 2 above)
Comment (srmukher) [reply packages/react-examples/src/react-charting/DonutChart/DonutChart.Multiple.Example.tsx]: done

File: packages/charts/react-charting/src/components/DonutChart/DonutChart.base.tsx (hunk 5 above)
Comment (srmukher) [reply packages/charts/react-charting/src/components/DonutChart/DonutChart.base.tsx]: separated the logic

File: packages/charts/react-charting/src/components/DonutChart/DonutChart.base.tsx [hunk 6]
```diff
@@ -265,11 +263,17 @@ export class DonutChartBase extends React.Component<IDonutChartProps, IDonutChar
         focusZonePropsInHoverCard={this.props.focusZonePropsForLegendsInHoverCard}
         overflowText={this.props.legendsOverflowText}
         {...this.props.legendProps}
+        canSelectMultipleLegends={this.props.canSelectMultipleLegends}
+        onChange={this._onLegendChange}
```
Comment (AtishayMsft) [inline packages/charts/react-charting/src/components/DonutChart/DonutChart.base.tsx:267]: >onChange={this._onLegendChange} [](http://example.com/codeflow?start=8&length=31)

keep using the action event handler. As this can be overridden for a specific legend as per the caller.

Dont modify existing behavior.

File: packages/charts/react-charting/src/components/DonutChart/DonutChart.types.ts [hunk 7]
```diff
@@ -69,6 +69,12 @@ export interface IDonutChartProps extends ICartesianChartProps {
    * @default false
    */
   roundCorners?: boolean;
+
+  /**
+   * Prop to enable the multiple legend selection
+   * @default false
+   */
+  canSelectMultipleLegends?: boolean;
```
Comment (AtishayMsft) [inline packages/charts/react-charting/src/components/DonutChart/DonutChart.types.ts:77]: >canSelectMultipleLegends?: boolean; [](http://example.com/codeflow?start=2&length=35)

not needed. This is an existing property of the legends control

File: packages/charts/react-charting/src/components/DonutChart/Arc/Arc.tsx [hunk 8]
```diff
@@ -31,12 +31,11 @@ export class Arc extends React.Component<IArcProps, IArcState> {
...
+    const opacity: number = activeArc && activeArc.length > 0 ? (isActive ? 1 : 0.1) : 1;
```
Comment (AtishayMsft) [inline packages/charts/react-charting/src/components/DonutChart/Arc/Arc.tsx:38]: >activeArc && activeArc.length > 0 ? (isActive ? 1 : 0.1) : 1; [](http://example.com/codeflow?start=28&length=61)

move this logic to isActive

File: packages/charts/react-charting/src/components/DonutChart/Arc/Arc.tsx [hunk 9]
```diff
@@ -129,7 +131,8 @@ export class Arc extends React.Component<IArcProps, IArcState> {
     if (
       hideLabels ||
       Math.abs(data!.endAngle - data!.startAngle) < Math.PI / 12 ||
-      (activeArc !== data!.data.legend && activeArc !== '')
+      !Array.isArray(activeArc) ||
+      (data!.data.legend && activeArc.includes(data!.data.legend))
```
Comment (AtishayMsft) [inline packages/charts/react-charting/src/components/DonutChart/Arc/Arc.tsx:135]: same logic is repeating

File: packages/charts/react-charting/src/components/DonutChart/DonutChart.base.tsx [hunk 10]
```diff
@@ -354,7 +365,7 @@ export class DonutChartBase extends React.Component<IDonutChartProps, IDonutChar
    * Note: This won't work in case of multiple legends selection.
    */
   private _getHighlightedLegend() {
```
Comment (AtishayMsft) [inline packages/charts/react-charting/src/components/DonutChart/DonutChart.base.tsx:367]: >_getHighlightedLegend [](http://example.com/codeflow?start=10&length=21)

modify this function for multiselect

File: packages/charts/react-charting/src/components/DonutChart/DonutChart.types.ts (hunk 7 above)
Comment (srmukher) [reply packages/charts/react-charting/src/components/DonutChart/DonutChart.types.ts]: done

File: packages/charts/react-charting/src/components/DonutChart/DonutChart.base.tsx (hunk 10 above)
Comment (srmukher) [reply packages/charts/react-charting/src/components/DonutChart/DonutChart.base.tsx:384]: this was already modified

File: packages/charts/react-charting/src/components/DonutChart/DonutChart.base.tsx (hunk 6 above)
Comment (srmukher) [reply packages/charts/react-charting/src/components/DonutChart/DonutChart.base.tsx:270]: modified as discussed offline

File: packages/charts/react-charting/src/components/DonutChart/Arc/Arc.tsx (hunk 8 above)
Comment (srmukher) [reply packages/charts/react-charting/src/components/DonutChart/Arc/Arc.tsx]: done

File: packages/charts/react-charting/src/components/DonutChart/Arc/Arc.tsx (hunk 9 above)
Comment (srmukher) [reply packages/charts/react-charting/src/components/DonutChart/Arc/Arc.tsx]: moved to a function

File: packages/charts/react-charting/etc/react-charting.api.md [hunk 11]
```diff
@@ -488,6 +488,7 @@ export interface IDonutChart {
 // @public
 export interface IDonutChartProps extends ICartesianChartProps {
     calloutProps?: Partial<ICalloutProps>;
+    canSelectMultipleLegends?: boolean;
```
Comment (AtishayMsft) [inline packages/charts/react-charting/etc/react-charting.api.md:491]: >canSelectMultipleLegends [](http://example.com/codeflow?start=4&length=24)

revert this change

Comment (srmukher) [reply packages/charts/react-charting/etc/react-charting.api.md]: done

File: packages/charts/react-charting/src/components/DonutChart/Arc/Arc.tsx [hunk 12]
```diff
@@ -123,13 +124,17 @@ export class Arc extends React.Component<IArcProps, IArcState> {
     return point.callOutAccessibilityData?.ariaLabel || (legend ? `${legend}, ` : '') + `${yValue}.`;
   };
 
+  private _shouldHighlightArc = (activeArc: string[], legend?: string): boolean => {
```
Comment (AtishayMsft) [inline packages/charts/react-charting/src/components/DonutChart/Arc/Arc.tsx:127]: >_shouldHighlightArc [](http://example.com/codeflow?start=10&length=19)

this same function can serve all 3 places.

File: packages/charts/react-charting/src/components/DonutChart/DonutChart.base.tsx [hunk 13]
```diff
@@ -24,9 +24,10 @@ export interface IDonutChartState {
   xCalloutValue?: string;
   yCalloutValue?: string;
   focusedArcId?: string;
-  selectedLegend: string;
+  selectedLegend?: string;
```
Comment (AtishayMsft) [inline packages/charts/react-charting/src/components/DonutChart/DonutChart.base.tsx:27]: >selectedLegend [](http://example.com/codeflow?start=2&length=14)

Selected legend is not being used anywhere

File: packages/charts/react-charting/src/components/DonutChart/DonutChart.base.tsx [hunk 14]
```diff
@@ -138,7 +140,9 @@ export class DonutChartBase extends React.Component<IDonutChartProps, IDonutChar
                 hoverLeaveCallback={this._hoverLeave}
                 uniqText={this._uniqText}
                 onBlurCallback={this._onBlur}
-                activeArc={this._getHighlightedLegend()}
+                activeArc={this._getHighlightedLegend().filter(
```
Comment (AtishayMsft) [inline packages/charts/react-charting/src/components/DonutChart/DonutChart.base.tsx:143]: >_getHighlightedLegend [](http://example.com/codeflow?start=32&length=21)

use isLegendHighlighted here

File: packages/charts/react-charting/src/components/DonutChart/DonutChart.base.tsx [hunk 15]
```diff
@@ -265,16 +266,41 @@ export class DonutChartBase extends React.Component<IDonutChartProps, IDonutChar
         focusZonePropsInHoverCard={this.props.focusZonePropsForLegendsInHoverCard}
         overflowText={this.props.legendsOverflowText}
         {...this.props.legendProps}
+        canSelectMultipleLegends={this.props.legendProps?.canSelectMultipleLegends}
```
Comment (AtishayMsft) [inline packages/charts/react-charting/src/components/DonutChart/DonutChart.base.tsx:269]: >canSelectMultipleLegends={this.props.legendProps?.canSelectMultipleLegends} [](http://example.com/codeflow?start=8&length=75)

this is already handled in previous line. Remove here

File: packages/charts/react-charting/src/components/DonutChart/DonutChart.base.tsx [hunk 16]
```diff
@@ -265,16 +266,41 @@ export class DonutChartBase extends React.Component<IDonutChartProps, IDonutChar
...
+        onChange={this._onLegendChange}
```
Comment (AtishayMsft) [inline packages/charts/react-charting/src/components/DonutChart/DonutChart.base.tsx:270]: >_onLegendChange [](http://example.com/codeflow?start=23&length=15)

onLegendSelectionChange

File: packages/charts/react-charting/src/components/DonutChart/DonutChart.base.tsx [hunk 17]
```diff
@@ -265,16 +266,41 @@ export class DonutChartBase extends React.Component<IDonutChartProps, IDonutChar
...
+    currentLegend?: ILegend,
+  ) => {
+    this.setState({ selectedLegends });
+    if (this.props.legendProps?.onChange) {
+      this.props.legendProps.onChange(selectedLegends, event, currentLegend);
+    }
+  };
+
+  private _isLegendHovered = (legend: string): boolean => {
+    return this._noLegendsHighlighted() || this._isLegendHighlighted(legend);
```
Comment (AtishayMsft) [inline packages/charts/react-charting/src/components/DonutChart/DonutChart.base.tsx:288]: >this._noLegendsHighlighted() [](http://example.com/codeflow?start=11&length=28)

this is redundant

File: packages/charts/react-charting/src/components/DonutChart/DonutChart.base.tsx [hunk 18]
```diff
@@ -354,7 +387,7 @@ export class DonutChartBase extends React.Component<IDonutChartProps, IDonutChar
    * Note: This won't work in case of multiple legends selection.
    */
   private _getHighlightedLegend() {
-    return this.state.selectedLegend || this.state.activeLegend;
+    return this.state.selectedLegends.length > 0 ? this.state.selectedLegends : [this.state.activeLegend];
```
Comment (AtishayMsft) [inline packages/charts/react-charting/src/components/DonutChart/DonutChart.base.tsx:390]: >return this.state.selectedLegends.length > 0 ? this.state.selectedLegends : [this.state.activeLegend]; [](http://example.com/codeflow?start=4&length=102)

this wont work if 2 legend are selected and a third legend is hovered

File: packages/charts/react-charting/src/components/DonutChart/DonutChart.base.tsx [hunk 19]
```diff
@@ -241,22 +245,19 @@ export class DonutChartBase extends React.Component<IDonutChartProps, IDonutChar
...
-          }
+          // Add any action logic here
         },
         hoverAction: () => {
           this._handleChartMouseLeave();
           this.setState({ activeLegend: point.legend! });
         },
         onMouseOutAction: () => {
-          this.setState({ activeLegend: '' });
+          this.setState({ activeLegend: undefined });
```
[the remark this answers is not in the document]
Comment (srmukher) [reply packages/charts/react-charting/src/components/DonutChart/DonutChart.base.tsx:255]: no, checked with a legend having '' title, it does not work in that case

File: packages/charts/react-charting/src/components/DonutChart/DonutChart.base.tsx (hunk 18 above)
Comment (srmukher) [reply packages/charts/react-charting/src/components/DonutChart/DonutChart.base.tsx:390]: But, in line multiple chart, on multiple legend selection, legend hover is disabled. Shall the experience be not the same?

File: packages/charts/react-charting/src/components/DonutChart/DonutChart.base.tsx (hunk 19 above)
[the remark this answers is not in the document]
Comment (AtishayMsft) [reply packages/charts/react-charting/src/components/DonutChart/DonutChart.base.tsx:255]: ok

File: packages/charts/react-charting/src/components/DonutChart/DonutChart.base.tsx (hunk 17 above)
Comment (srmukher) [reply packages/charts/react-charting/src/components/DonutChart/DonutChart.base.tsx:288]: keeping this for now as we will take up the change in experience task later.

File: packages/charts/react-charting/src/components/DonutChart/DonutChart.base.tsx (hunk 16 above)
Comment (srmukher) [reply packages/charts/react-charting/src/components/DonutChart/DonutChart.base.tsx:270]: done

File: packages/charts/react-charting/src/components/DonutChart/DonutChart.base.tsx (hunk 15 above)
Comment (srmukher) [reply packages/charts/react-charting/src/components/DonutChart/DonutChart.base.tsx:269]: done

File: packages/charts/react-charting/src/components/DonutChart/DonutChart.base.tsx [hunk 20]
```diff
@@ -241,22 +245,19 @@ export class DonutChartBase extends React.Component<IDonutChartProps, IDonutChar
         title: point.legend!,
         color,
         action: () => {
```
[the remark this answers is not in the document]
Comment (srmukher) [reply packages/charts/react-charting/src/components/DonutChart/DonutChart.base.tsx:247]: done

File: packages/charts/react-charting/src/components/DonutChart/DonutChart.base.tsx (hunk 13 above)
Comment (srmukher) [reply packages/charts/react-charting/src/components/DonutChart/DonutChart.base.tsx:27]: removed

File: packages/charts/react-charting/src/components/DonutChart/DonutChart.base.tsx [hunk 21]
```diff
@@ -265,16 +259,41 @@ export class DonutChartBase extends React.Component<IDonutChartProps, IDonutChar
         focusZonePropsInHoverCard={this.props.focusZonePropsForLegendsInHoverCard}
         overflowText={this.props.legendsOverflowText}
         {...this.props.legendProps}
+        // eslint-disable-next-line react/jsx-no-bind
+        onChange={this._onLegendSelectionChange.bind(this)}
```
Comment (AtishayMsft) [inline packages/charts/react-charting/src/components/DonutChart/DonutChart.base.tsx:263]: >this._onLegendSelectionChange.bind(this) [](http://example.com/codeflow?start=18&length=40)

do you need to use this twice?

File: packages/charts/react-charting/src/components/DonutChart/DonutChart.base.tsx [hunk 22]
```diff
@@ -265,16 +259,41 @@ export class DonutChartBase extends React.Component<IDonutChartProps, IDonutChar
...
 
+  private _onLegendSelectionChange(
+    selectedLegends: string[],
+    event: React.MouseEvent<HTMLButtonElement>,
+    currentLegend?: ILegend,
+  ): void {
+    if (this.props.legendProps?.canSelectMultipleLegends) {
+      this.setState({ selectedLegends });
+    } else {
+      this.setState({ selectedLegends: selectedLegends.slice(-1) });
```
Comment (AtishayMsft) [inline packages/charts/react-charting/src/components/DonutChart/DonutChart.base.tsx:277]: >selectedLegends.slice(-1) [](http://example.com/codeflow?start=39&length=25)

does this work when the legend is unselected

File: packages/charts/react-charting/src/components/DonutChart/DonutChart.base.tsx [hunk 23]
```diff
@@ -265,16 +259,41 @@ export class DonutChartBase extends React.Component<IDonutChartProps, IDonutChar
...
+    }
+  }
+
+  private _isLegendHighlighted = (legend: string): boolean => {
+    return this._getHighlightedLegend().indexOf(legend) > -1;
+  };
+
+  private _noLegendsHighlighted = (): boolean => {
+    return this._getHighlightedLegend().length === 0;
+  };
```
Comment (AtishayMsft) [inline packages/charts/react-charting/src/components/DonutChart/DonutChart.base.tsx:290]: keep these functions together with getHighlightedLegends

File: packages/charts/react-charting/src/components/DonutChart/DonutChart.base.tsx [hunk 24]
```diff
@@ -265,16 +259,41 @@ export class DonutChartBase extends React.Component<IDonutChartProps, IDonutChar
...
+
   private _focusCallback = (data: IChartDataPoint, id: string, element: SVGPathElement): void => {
     this._currentHoverElement = element;
     this.setState({
       /** Show the callout if highlighted arc is focused and Hide it if unhighlighted arc is focused */
-      showHover: this.state.selectedLegend === '' || this.state.selectedLegend === data.legend,
+      showHover: this._noLegendsHighlighted() || this._isLegendHighlighted(data.legend!),
```
Comment (AtishayMsft) [inline packages/charts/react-charting/src/components/DonutChart/DonutChart.base.tsx:296]: >! [](http://example.com/codeflow?start=86&length=1)

move this ! check inside isLegendHighlighted function

File: packages/charts/react-charting/src/components/DonutChart/DonutChart.base.tsx [hunk 25]
```diff
@@ -324,16 +343,23 @@ export class DonutChartBase extends React.Component<IDonutChartProps, IDonutChar
...
-    const highlightedLegend = this._getHighlightedLegend();
-    if (valueInsideDonut !== undefined && (highlightedLegend !== '' || this.state.showHover)) {
-      let legendValue = valueInsideDonut;
-      data!.map((point: IChartDataPoint, index: number) => {
-        if (point.legend === highlightedLegend || (this.state.showHover && point.legend === this.state.legend)) {
-          legendValue = point.yAxisCalloutData ? point.yAxisCalloutData : point.data!;
+    const highlightedLegends = this._getHighlightedLegend().filter(legend => legend !== undefined);
+
+    if (this.state.showHover) {
+      const legendValue = data.find(point => point.legend === this.state.legend);
```
Comment (AtishayMsft) [inline packages/charts/react-charting/src/components/DonutChart/DonutChart.base.tsx:349]: >legendValue [](http://example.com/codeflow?start=12&length=11)

call this variable point

File: packages/charts/react-charting/src/components/DonutChart/DonutChart.base.tsx [hunk 26]
```diff
@@ -265,16 +259,41 @@ export class DonutChartBase extends React.Component<IDonutChartProps, IDonutChar
...
+    } else {
+      this.setState({ selectedLegends: selectedLegends.slice(-1) });
+    }
+    if (this.props.legendProps?.onChange) {
+      this.props.legendProps.onChange(selectedLegends, event, currentLegend);
+    }
+  }
+
+  private _isLegendHighlighted = (legend: string): boolean => {
+    return this._getHighlightedLegend().indexOf(legend) > -1;
```
Comment (AtishayMsft) [inline packages/charts/react-charting/src/components/DonutChart/DonutChart.base.tsx:285]: >.indexOf(legend) > -1; [](http://example.com/codeflow?start=39&length=22)

use .includes?

File: packages/charts/react-charting/src/components/DonutChart/Arc/Arc.tsx [hunk 27]
```diff
@@ -123,13 +124,16 @@ export class Arc extends React.Component<IArcProps, IArcState> {
...
```
Comment (AtishayMsft) [inline packages/charts/react-charting/src/components/DonutChart/Arc/Arc.tsx:127]: >activeArc [](http://example.com/codeflow?start=33&length=9)

this can from this.props. Not needed as argument. All the empty and length checks can move to this. Right?

File: packages/charts/react-charting/src/components/DonutChart/__snapshots__/DonutChart.test.tsx.snap [hunk 28]
```diff
@@ -2443,22 +2429,6 @@ exports[`DonutChart snapShot testing Should render arc labels in percentage form
               onMouseOver={[Function]}
               role="img"
             />
-            <text
```
Comment (AtishayMsft) [inline packages/charts/react-charting/src/components/DonutChart/__snapshots__/DonutChart.test.tsx.snap:2446]: >text [](http://example.com/codeflow?start=13&length=4)

why this change

File: packages/charts/react-charting/src/components/DonutChart/__snapshots__/DonutChartRTL.test.tsx.snap [hunk 29]
```diff
@@ -636,498 +636,4 @@ exports[`Donut chart interactions Should hide callout on mouse leave 1`] = `
...
-              </div>
-            </div>
-          </div>
-        </div>
-      </div>
-    </div>
-  </div>
-</div>
-`;
+exports[`Donut chart interactions Should reflect theme change 1`] = `<div />`;
```
Comment (AtishayMsft) [inline packages/charts/react-charting/src/components/DonutChart/__snapshots__/DonutChartRTL.test.tsx.snap:639]: >`<div />`; [](http://example.com/codeflow?start=68&length=10)

why did the div disappear

File: packages/charts/react-charting/src/components/DonutChart/DonutChart.base.tsx (hunk 22 above)
Comment (srmukher) [reply packages/charts/react-charting/src/components/DonutChart/DonutChart.base.tsx:277]: yes working

File: packages/charts/react-charting/src/components/DonutChart/DonutChart.base.tsx (hunk 26 above)
Comment (srmukher) [reply packages/charts/react-charting/src/components/DonutChart/DonutChart.base.tsx:285]: done

File: packages/charts/react-charting/src/components/DonutChart/DonutChart.base.tsx (hunk 24 above)
Comment (srmukher) [reply packages/charts/react-charting/src/components/DonutChart/DonutChart.base.tsx:296]: done

File: packages/charts/react-charting/src/components/DonutChart/DonutChart.base.tsx (hunk 25 above)
Comment (srmukher) [reply packages/charts/react-charting/src/components/DonutChart/DonutChart.base.tsx:347]: calling it pointValue as point is used in data.find

File: packages/charts/react-charting/src/components/DonutChart/DonutChart.base.tsx (hunk 23 above)
Comment (srmukher) [reply packages/charts/react-charting/src/components/DonutChart/DonutChart.base.tsx:289]: done

File: packages/charts/react-charting/src/components/DonutChart/DonutChart.base.tsx (hunk 21 above)
Comment (srmukher) [reply packages/charts/react-charting/src/components/DonutChart/DonutChart.base.tsx:262]: This is used just once

File: packages/charts/react-charting/src/components/DonutChart/Arc/Arc.tsx (hunk 27 above)
Comment (srmukher) [reply packages/charts/react-charting/src/components/DonutChart/Arc/Arc.tsx]: moved

File: packages/charts/react-charting/src/components/DonutChart/__snapshots__/DonutChart.test.tsx.snap (hunk 28 above)
Comment (srmukher) [reply packages/charts/react-charting/src/components/DonutChart/__snapshots__/DonutChart.test.tsx.snap]: updated snapshots

File: packages/charts/react-charting/src/components/DonutChart/__snapshots__/DonutChartRTL.test.tsx.snap (hunk 29 above)
Comment (srmukher) [reply packages/charts/react-charting/src/components/DonutChart/__snapshots__/DonutChartRTL.test.tsx.snap]: updated snapshots

File: packages/charts/react-charting/src/components/DonutChart/Arc/Arc.tsx [hunk 30]
```diff
@@ -123,13 +124,17 @@ export class Arc extends React.Component<IArcProps, IArcState> {
     return point.callOutAccessibilityData?.ariaLabel || (legend ? `${legend}, ` : '') + `${yValue}.`;
   };
 
-  private _renderArcLabel = (className: string) => {
-    const { arc, data, innerRadius, outerRadius, showLabelsInPercent, totalValue, hideLabels, activeArc } = this.props;
+  private _shouldHighlightArc = (legend?: string): boolean => {
+    const { activeArc } = this.props;
+    return !(activeArc && activeArc.length > 0 && legend !== undefined && !activeArc.includes(legend));
```
Comment (AtishayMsft) [inline packages/charts/react-charting/src/components/DonutChart/Arc/Arc.tsx:129]: >! [](http://example.com/codeflow?start=11&length=1)

why is there negation sign here

File: packages/charts/react-charting/src/components/DeclarativeChart/PlotlySchemaAdapter.ts [hunk 31]
```diff
@@ -94,6 +94,9 @@ export const transformPlotlyJsonToDonutProps = (
     hideLabels,
     showLabelsInPercent: firstData.textinfo ? firstData.textinfo === 'percent' : true,
     styles,
+    legendProps: {
+      canSelectMultipleLegends: true,
+    },
```
Comment (AtishayMsft) [inline packages/charts/react-charting/src/components/DeclarativeChart/PlotlySchemaAdapter.ts:99]: legend props are defined in DeclarativeChart.tsx. Use that only.

File: packages/charts/react-charting/src/components/DonutChart/DonutChart.base.tsx [hunk 32]
```diff
@@ -25,9 +25,9 @@ export interface IDonutChartState {
   xCalloutValue?: string;
   yCalloutValue?: string;
   focusedArcId?: string;
-  selectedLegend: string;
   dataPointCalloutProps?: IChartDataPoint;
   callOutAccessibilityData?: IAccessibilityProps;
+  selectedLegends: string[];
```
Comment (AtishayMsft) [inline packages/charts/react-charting/src/components/DonutChart/DonutChart.base.tsx:30]: >selectedLegends [](http://example.com/codeflow?start=2&length=15)

can you resolve all addressed comments

File: packages/charts/react-charting/src/components/DonutChart/DonutChartRTL.test.tsx [hunk 33]
```diff
@@ -184,6 +184,38 @@ describe('Donut chart interactions', () => {
...
+    fireEvent.click(firstLegend!);
+    fireEvent.click(secondLegend!);
+
+    // Assert
+    expect(firstLegend).toHaveAttribute('aria-selected', 'true');
+    expect(secondLegend).toHaveAttribute('aria-selected', 'true');
+
+    const getById = queryAllByAttribute.bind(null, 'id');
+    expect(getById(container, /Pie.*?first/i)[0]).toHaveStyle('opacity: 1.0');
+    expect(getById(container, /Pie.*?second/i)[0]).toHaveStyle('opacity: 1.0');
```
Comment (AtishayMsft) [inline packages/charts/react-charting/src/components/DonutChart/DonutChartRTL.test.tsx:216]: this is good. Thanks.

File: packages/charts/react-charting/src/components/DonutChart/DonutChart.base.tsx (hunk 21 above)
Comment (srmukher) [reply packages/charts/react-charting/src/components/DonutChart/DonutChart.base.tsx:270]: yes, if we don't use .bind(this), tests are failing

File: packages/charts/react-charting/src/components/DeclarativeChart/PlotlySchemaAdapter.ts (hunk 31 above)
Comment (srmukher) [reply packages/charts/react-charting/src/components/DeclarativeChart/PlotlySchemaAdapter.ts:99]: done

File: packages/charts/react-charting/src/components/DonutChart/Arc/Arc.tsx (hunk 30 above)
Comment (srmukher) [reply packages/charts/react-charting/src/components/DonutChart/Arc/Arc.tsx]: simplified the expression

**Review by AtishayMsft — APPROVED** [review]
